src/interface/misc.rs

```diff
@@ -381,70 +381,58 @@ pub struct RustSemaphore {
 impl RustSemaphore {
     pub fn new(value_handle: u32, is_shared: bool) -> Self {
         Self {
-            value: RustAtomicU32::new(value_handle),
+            value: Mutex::new(value_handle),
             is_shared: RustAtomicBool::new(is_shared),
         }
     }
 
-    pub fn lock(&self) -> bool{
-        while self.value.load(RustAtomicOrdering::Relaxed) == 0 { interface::lind_yield(); }
-
-        let result = self.value.fetch_update(RustAtomicOrdering::Relaxed, RustAtomicOrdering::Relaxed, |x| {
-            if x > 0 { Some(x - 1) } else { Some(0) }
-        });
-
-        match result {
-            Ok(_) => true,
-            Err(_) => false,
+    pub fn lock(&self) {
+        loop {
+            let mut value = self.value.lock();
+            if *value == 0 {
+                interface::lind_yield();
+                continue;
+            }
+                
+            *value = if *value > 0 { *value - 1 } else { 0 };
+            break;
         }
     }
 
-    pub fn unlock(&self) -> bool {
-        let result = self.value.fetch_update(RustAtomicOrdering::Relaxed, RustAtomicOrdering::Relaxed, |x| {
-            if x < (SEM_VALUE_MAX - 1) { Some(x + 1) } else { Some(SEM_VALUE_MAX) }
-        });
-
-        match result {
-            Ok(_) => true,
-            Err(_) => false,
-        }
-
+    pub fn unlock(&self) {
+        let mut value = self.value.lock();
+        *value = if *value < (SEM_VALUE_MAX - 1) { *value + 1 } else { SEM_VALUE_MAX };
```

Comment:
According to the [sem_post](https://man7.org/linux/man-pages/man3/sem_post.3.html) man page, we may need to report error=EOVERFLOW when the value larger than SEM_VALUE_MAX?

Reply:
Good catch we probably need to return an error here

Reply:
Thanks for pointing this out. I've made the relevant changes.